### PR TITLE
New feature: inactive top cells

### PR DESCRIPTION
**Pull request by cbegeman**
Inactive top cells are introduced to add flexibility to the vertical coordinate below ice shelves. This PR introduces phase 1 in the development documented in https://github.com/MPAS-Dev/MPAS-Model/pull/811, which touches only the code needed to run a basic ice shelf test case.

This is a work-in-progress.

### CODE REVIEW

**Comment by cbegeman** [comment]
This is a first stab at dividing up the work. Feel free to edit or suggest changes if what I've suggested doesn't suit. I roughly ordered the list by priority, where we make the changes to the parts of the code that cannot be disabled first. This basically means that the tendency terms are updated later. For the most part, there's only one person touching each module and one person for each of the tendency groups (velocity, tracers, thickness). 

- "[X] E3SM" indicates that the changes in the module/subroutine passed E3SM tests (details TBD)
- "[X] Standalone" indicates that the changes in the module/subroutine passed the baseline regression suite for MPAS-Ocean [here](https://github.com/xylar/compass/tree/remove_config_iterative_init_variable)
- The checkbox at the start of the line indicates that the changes have been incorporated into this PR. Ideally this wouldn't happen until after it passes the corresponding standalone and E3SM tests. Right now there are a few exceptions to this rule to get us started.

Note: `minLevel*` variables are available through `use ocn_mesh`.
Note: use this loop syntax for now (Option 1 in [this discussion](https://github.com/MPAS-Dev/MPAS-Model/pull/811#issuecomment-777691337)):
```
do k = minLevelCell(iEdge), maxLevelCell(iEdge) !mlc-do-k
end do !mlc-end-do
```
Note: This may still not be a complete list. 

Init:
- [X] ocn_init_sub_ice_shelf_2D: @xylar [X] E3SM
   - [X] ocn_init_setup_sub_ice_shelf_2D() [X] Standalone
- [X] ocn_init_ssh_and_landIcePressure: @xylar
   - [X] ocn_init_ssh_and_landIcePressure_balance() [X] Standalone
- [x] ocn_init_vertical_grids: @xylar [x] Standalone [X] E3SM
   - [X] ocn_init_vertical_grid() [X] Standalone
   - [x] ocn_compute_z_level_layerThickness()  [X] Standalone
   - [x] ocn_compute_zMid_from_layerThickness()  [X] Standalone
   - [x] ocn_compute_z_star_layerThickness()  [X] Standalone
   - [x] ocn_alter_bottomDepth_for_pbcs()  [X] Standalone
   - [x] ocn_alter_ssh_for_ptcs()  [X] Standalone
   - [x] ocn_compute_Haney_number() [X] Standalone
   - [ ] ~~ocn_init_vertical_grid_with_max_rx1() [ ] Standalone~~  Probably doesn't make sense with `minLevelCell > 1`
   - [ ] ~~constrain_rx1_layer() [ ] Standalone~~ Same as above

Forward:
- [X] ocn_init_routines: @cbegeman [X] E3SM
   - [X] ocn_init_routines_compute_min_max_level() [X] Standalone
   - [X] ocn_init_routines_vert_coord() [X] Standalone
   - [X] ocn_init_routines_block() [X] Standalone
- [X] ocn_forcing: @xylar [X] E3SM
   - [X] ocn_forcing_build_fraction_absorbed_array [X] Standalone
- [X] ocn_mesh: @cbegeman [ ] Standalone [ ] E3SM
   - [X] minLevelCell
   - [X] minLevelEdgeTop
   - [X] minLevelEdgeBot
   - [X] minLevelVertexTop
   - [X] minLevelVertexBot
   - [X] ocn_meshCreate()
   - [X] ocn_meshUpdateFields()
   - [X] ocn_meshDestroy()
- [X] ocn_time_integration_split: @cbegeman [X] E3SM
   - [X] ocn_time_integration_split_init()  [X] Standalone
   - [X] ocn_time_integrator_split()  [X] Standalone
      - [X] Outside large iteration loop  [X] Standalone
      - [X] Inside large iteration loop  [X] Standalone
         Note: nothing is done here when config_disable_thick_all_tend .and. config_disable_vel_all_tend .and. config_disable_tr_all_tend
- [X] ocn_diagnostics: @cbegeman [X] Standalone [X] E3SM
   - [X] ocn_relativeVorticity_circulation() [X] Standalone
   - [X] ocn_diagnostic_solve_layerThicknessEdge() [X] Standalone
   - [X] ocn_diagnostic_solve_vorticity()
   - [X] ocn_diagnostic_solve_richardson() [X] Standalone
   - [X] ocn_diagnostic_solve_surfaceLayer() [X] Standalone
   - [X] ocn_diagnostic_solve_vortVel() [X] Standalone
   - [X] ocn_diagnostic_solve_z_coordinates() [X] Standalone
   - [X] ocn_diagnostic_solve_pressure() [X] Standalone
   - [X] ocn_vert_transport_velocity_top()
   - [X] ocn_fuperp() [X] Standalone
   - [X] ocn_compute_land_ice_flux_input_fields()
   - [X] ocn_validate_state() [X] Standalone
- [X] ocn_thick_ale: @cbegeman [X] Standalone [X] E3SM
   - [X] ocn_ALE_thickness
- [X] ocn_vmix_cvmix: @cbegeman [X] Standalone [X] E3SM
   - [X] ocn_vmix_coefs_cvmix_build	 

- [x] ocn_thick_hadv: @xylar [X] Standalone [X] E3SM
   - [x] ocn_hadv_thick_tend (config_disable_thick_hadv)
- [x] ocn_thick_vadv: @xylar [X] Standalone [X] E3SM
   - [x] ocn_vadv_thick_tend (config_disable_thick_vadv)
- [x] ocn_thick_surface_flux: @xylar [X] Standalone [X] E3SM
   - [x] ocn_thick_surface_flux_tend (config_disable_thick_sflux)	 

- [ ] ~~ocn_tendency~~: @cbegeman
   - [ ] ~~ocn_tend_tracer~~ (config_disable_tr_all_tend) Belongs in Phase 2

- [x] ocn_tracer_advection: @xylar [X] Standalone [X] E3SM
   - [x] ocn_tracer_advection_tend (config_disable_tr_adv)
- [x] ocn_tracer_advection_mono: @xylar [X] Standalone  [X] E3SM
   - [x] ocn_tracer_advection_mono_tend (config_disable_tr_adv)

- [x] ocn_vel_forcing_surface_stress: @xylar  [X] Standalone [X] E3SM
   - [x] ocn_vel_forcing_surface_stress_tend (config_disable_vel_surface_stress)
- [x] ocn_vel_hadv_coriolis: @xylar  [X] Standalone [X] E3SM
   - [x] ocn_vel_hadv_coriolis_tend (config_disable_vel_coriolis)
- [x] ocn_vel_vadv: @xylar  [X] Standalone [X] E3SM
   - [x] ocn_vel_vadv_tend (config_disable_vel_vadv)
- [x] ocn_vel_pressure_grad: @xylar  [X] Standalone [X] E3SM
   - [x] ocn_vel_pressure_grad_tend (config_disable_vel_pgrad)

- [X] ocn_vmix: @cbegeman [X] Standalone [X] E3SM
   - [X] ocn_vmix_implicit
   - [X] ocn_vel_vmix_tend_implicit (config_disable_vel_vmix)
   - [X] ocn_tracer_vmix_tend_implicit (config_disable_tr_vmix)

- [X] ocn_vel_hmix_del2: @cbegeman [X] Standalone [X] E3SM
   - [X] ocn_vel_hmix_del2_tend (config_disable_vel_hmix)

File: src/core_ocean/Registry.xml [hunk 1]
```diff
@@ -1629,6 +1631,7 @@
 			<var name="layerThickness"/>
 			<var name="ssh"/>
 			<var name="maxLevelEdgeTop"/>
+			<var name="minLevelEdgeTop"/>
```
Comment (xylar) [inline src/core_ocean/Registry.xml]: I think `minLevelEdgeBot` would be the corresponding variable to `maxLevelEdgeTop`.

File: src/core_ocean/shared/mpas_ocn_init_routines.F [hunk 2]
```diff
@@ -502,24 +553,27 @@ subroutine ocn_init_routines_vert_coord(domain)!{{{
...
+                     bottomDepth(iCell))>20.0_RKIND) then
                     consistentSSH = .false.
                     call mpas_log_write(' Warning: Sea surface height is outside of acceptable physical range, i.e. abs(sum(h)-bottomDepth)>20m.', &
                        MPAS_LOG_ERR)
-                    call mpas_log_write(' iCell: $i, maxLevelCell(iCell): $i, bottomDepth(iCell): $r, sum(h): $r', &
-                       intArgs=(/iCell, maxLevelCell(iCell) /), &
-                       realArgs=(/ bottomDepth(iCell),sum(layerThickness(1:maxLevelCell(iCell),iCell)) /) )
+                    call mpas_log_write(' iCell: $i, minLevelCell(iCell), maxLevelCell(iCell): $i, bottomDepth(iCell): $r, sum(h): $r', &
+                       intArgs=(/iCell, minLevelCell(iCell), maxLevelCell(iCell) /), &
+                       realArgs=(/ bottomDepth(iCell),sum(layerThickness(minLevelCell(iCell):maxLevelCell(iCell),iCell)) /) )
```
Comment (xylar) [inline src/core_ocean/shared/mpas_ocn_init_routines.F]:
```suggestion
                    call mpas_log_write(' iCell: $i, minLevelCell(iCell), maxLevelCell(iCell): $i, $i, bottomDepth(iCell): $r, sum(h): $r', &
                       intArgs=(/iCell, minLevelCell(iCell), maxLevelCell(iCell) /), &
                       realArgs=(/ bottomDepth(iCell),sum(layerThickness(minLevelCell(iCell):maxLevelCell(iCell),iCell)) /) )
```

File: src/core_ocean/shared/mpas_ocn_init_routines.F [hunk 3]
```diff
@@ -502,24 +553,27 @@ subroutine ocn_init_routines_vert_coord(domain)!{{{
...
```
Comment (xylar) [inline src/core_ocean/shared/mpas_ocn_init_routines.F]:
```suggestion
                    call mpas_log_write(' iCell: $i, minLevelCell(iCell), maxLevelCell(iCell): $i, $i bottomDepth(iCell): $r, sum(h): $r', &
                       intArgs=(/iCell, minLevelCell(iCell), maxLevelCell(iCell) /), &
                       realArgs=(/ bottomDepth(iCell),sum(layerThickness(minLevelCell(iCell):maxLevelCell(iCell),iCell)) /) )
```

File: src/core_ocean/shared/mpas_ocn_mesh.F [hunk 4]
```diff
@@ -307,7 +312,17 @@ subroutine ocn_meshCreate(domain) !{{{
...
+                                  minLevelCell)
+         call mpas_pool_get_array(meshPool, 'minLevelEdgeTop', &
+                                  minLevelEdgeTop)
+         call mpas_pool_get_array(meshPool, 'minLevelEdgeBot', &
+                                  minLevelEdgeBot)
+         call mpas_pool_get_array(meshPool, 'minLevelVertexTop', &
+                                  minLevelVertexTop)
+         call mpas_pool_get_array(meshPool, 'minLevelVertexBot', &
+                                  minLevelVertexBot)
+         call mpas_pool_get_array(meshPool, 'minLevelCell', &
```
Comment (xylar) [inline src/core_ocean/shared/mpas_ocn_mesh.F]:
```suggestion
         call mpas_pool_get_array(meshPool, 'maxLevelCell', &
```

**Comment by xylar** [comment]
@cbegeman, I found a few small things to correct in your changes.  Nothing that should affect us proceeding but things that need to be fixed soonish.

Also, you will need to add @dengwirda and me as collaborators (with write or maybe triage permission) to your fork so we can update the check marks above.  Let me know if you want more instructions on how to do that.

File: src/core_ocean/shared/mpas_ocn_init_routines.F [hunk 5]
```diff
@@ -93,7 +93,7 @@ module ocn_init_routines
 !>  This routine initializes max level and boundary mask variables
 !
 !-----------------------------------------------------------------------
-subroutine ocn_init_routines_compute_max_level(domain)!{{{
+subroutine ocn_init_routines_compute_min_max_level(domain)!{{{
```
Comment (xylar) [inline src/core_ocean/shared/mpas_ocn_init_routines.F:96]: The corresponding `end subroutine ocn_init_routines_compute_min_max_level` change is needed.

**Comment by xylar** [comment]
The branch as it is doesn't compile for me because of the missing change above (`end subroutine ocn_init_routines_compute_min_max_level`) so that's a high priority to fix.

**Comment by xylar** [comment]
There are 2 other places where the update `ocn_init_routines_compute_max_level` --> `ocn_init_routines_compute_min_max_level` is needed:
https://github.com/cbegeman/MPAS-Model/blob/ocean/develop/src/core_ocean/mode_analysis/mpas_ocn_analysis_mode.F#L115
and
https://github.com/cbegeman/MPAS-Model/blob/ocean/develop/src/core_ocean/shared/mpas_ocn_init_routines.F#L68

It should also be fixed in this comment:
https://github.com/cbegeman/MPAS-Model/blob/ocean/develop/src/core_ocean/shared/mpas_ocn_init_routines.F#L87

**Comment by xylar** [comment]
The fix I list above in `ocn_meshCreate` is also needed in order for the COMPASS regression tests to pass.

**Comment by cbegeman** [comment]
@xylar Thanks for catching those mistakes. I think I've made all the changes you've requested. I'll get in touch about testing it with the regression suite.

**Comment by xylar** [comment]
I added test cases based on @mark-petersen's comment: https://github.com/MPAS-Dev/MPAS-Model/pull/811#issuecomment-780079421
If you check out (or fetch and hard reset) my compass branch https://github.com/xylar/compass/tree/add_z_level_sub_ice_shelf_2D
```
$ ./list_testcases.py | grep z_level
  190: -o ocean -c sub_ice_shelf_2D -r 5km -t z_level
  191: -o ocean -c sub_ice_shelf_2D -r 5km -t z_level_full_cells
  192: -o ocean -c sub_ice_shelf_2D -r 5km -t z_level_full_cells_const_S
```
There are 3 versions of the `z_level*` test case that are hopefully self-explanatory.  I have also disabled all tendencies by default for now in the `forward` run.

I also updated this branch because I had made a mistake in the definition of full top cells.  I'll add my commit for adding `minLevelCell` to `ocn_compute_Haney_number` in a moment, too.

One way to prioritize our work on the modules and subroutines would be to find out where the `z_level_full_cells_const_S/forward` step breaks.  I'll play with this now.

**Comment by xylar** [comment]
Current error I'm seeing is:
```
Program received signal SIGFPE: Floating-point exception - erroneous arithmetic operation.

Backtrace for this error:
#0  0x7f3c526b020f in ???
#1  0x55f80b1e4404 in __ocn_time_integration_split_MOD_ocn_time_integration_split_init
	at /home/xylar/code/mpas-work/compass/add_z_level_sub_ice_shelf_2D/MPAS-Model/ocean/new-min-level-cell-feature/src/core_ocean/mode_forward/mpas_ocn_time_integration_split.F:2156
```
I saw that @cbegeman has `ocn_time_integration_split_init` checked off but I guess that commit hasn't been pushed to the branch yet?

**Comment by xylar** [comment]
Regarding `ocn_init_vertical_grids`:

I won't be putting `minLevelCell` in 2 functions related to the vertical coordinate with Haney-number constraint.  That would be a huge can of worms and would take quite a bit of thinking.  For now, the only vertical coordinate that supports `minLevelCell > 1` is `z-level`, not the `haney-number` coordinate created by this code.  So I think it's fine.

## Standalone testing

I tested on my Ubuntu laptop with Gnu and on Anvil with Intel, and the `z_level_dev` regression suite passed against the baseline.  For reference, the baseline is with compass branch:
https://github.com/xylar/compass/tree/remove_config_iterative_init_variable
and MPAS-Model branch:
https://github.com/xylar/MPAS-Model/tree/ocean/simplify_ssh_land_ice_pressure_init

## E3SM testing

I ran:
```
PEM_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.anvil_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel
PET_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.anvil_gnu
PET_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.anvil_gnu
PET_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
PET_Ln9.T62_oQU240.GMPAS-IAF.anvil_intel
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_gnu
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel
```
The last 2 were compared against a baseline of today's `E3SM/master`.

I will not be recording my testing at this level of detail in the future, just differences from this protocol.

**Comment by xylar** [comment]
## testing of `ocn_forcing`

### COMPASS

The `z_level_dev` regression test passed with Gnu compilers on my Ubuntu laptop and Intel on Anvil, with the baseline listed in my previous post.

### E3SM

The following tests passed:
```
PEM_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.anvil_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
PET_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.anvil_gnu
PET_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.anvil_gnu
PET_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_gnu.G  # baseline from today's E3SM/master
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_gnu.C
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel.G  # baseline from today's E3SM/master
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel.C
```

The following test failed because of a missing data file unrelated to this work (see https://github.com/E3SM-Project/E3SM/pull/4092#issuecomment-782009399):
```
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel
```

**Comment by xylar** [comment]
```
do k = minLevelCell(iEdge), maxLevelCell(iEdge) !mlc-do-k
end do !mlc-end-do
```
My understanding from [Phil's comment](https://github.com/MPAS-Dev/MPAS-Model/pull/811#issuecomment-781464654)  on the design document is that he's saying we don't need these comments.  @cbegeman, if that's your reading of that comment, too, I suggest we stop including these from now on and maybe remove the existing ones before making the "real" PR.

**Comment by xylar** [comment]
## testing of `ocn_thick_hadv`, `*_vadv` and `*_surface_flux`

### COMPASS

The `z_level_dev` regression test passed with Gnu compilers on my Ubuntu laptop and Intel on Anvil, with the baseline listed in my previous posts.

### E3SM

The following tests passed:
```
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel
PEM_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.anvil_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
PET_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.anvil_gnu
PET_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.anvil_gnu
PET_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_gnu.C
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel.C
```

@jonbob staged the missing file for the first test, so it is working again!

**Comment by xylar** [comment]
@cbegeman, please assign me some more modules to work on unless there aren't any.

**Comment by cbegeman** [comment]
### Testing of ocn_init_routines, *_validate_state, *_time_integration_split

**COMPASS**

The z_level_dev regression test passed with Intel compilers on grizzly against the baseline.

Baseline: 
MPAS-Dev/compass/legacy, https://github.com/MPAS-Dev/compass/commit/9f3378b278c3afd1c64c261c772271575ac426fe
with cherry-picked 
`commit 79a2aa46143f4cbecc735ffd143a1c0fd72f1567  REMOVE LATER!!! Add z-leve dev. test suite`
from https://github.com/xylar/compass/tree/add_z_level_sub_ice_shelf_2D


**E3SM**

The following tests passed:
```
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_intel.C
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu.C
PET_Ln9.T62_oQU240.GMPAS-IAF.cori-knl_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.cori-haswell_gnu
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-haswell_intel
PET_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.cori-knl_gnu
PEM_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.cori-knl_intel
```

**Comment by xylar** [comment]
@cbegeman, I noticed you force-pushed to this branch.  If we're sharing the branch, that's not safe to do.  It could mean changes get lost.  If you want to edit or clean up the branch history, we should switch to having me (and @dengwirda) do PRs to this branch instead.

**Comment by xylar** [comment]
I just put back my 3 commits that got force-removed.

**Comment by xylar** [comment]
## testing of `ocn_vel_hadv_coriolis`, `*_vadv`, `*_forcing_surface_stress`, and `*_pressure_grad`

### COMPASS

The `z_level_dev` regression test passed with Gnu compilers on my Ubuntu laptop and Intel on Anvil, with the baseline listed in my previous posts.

### E3SM

The following tests passed:
```
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_gnu.C
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel.C
PEM_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
PEM_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.anvil_intel
PET_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
```

But these failed:
```
PET_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.anvil_gnu
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel
```
I'm working on rolling things back to figure out why.

**Update:**
```
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel
```
is failing on `E3SM/master`.  I did a fresh clone to make sure.  Results are here:
```
/lcrc/group/e3sm/ac.xylar/scratch/anvil/PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel.20210220_061822_7tihl6
```
but
```
PET_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.anvil_gnu
```
is succeeding on `E3SM/master`, and failing on this branch so that one needs investigation.
This test is succeeding on 13c14fb1 but failing on a8cbcf69.  This is very odd because I'm quite sure the latter was part of my testing yesterday, and it succeeded.
I tried again with a8cbcf69 and it passed this time.  I really don't get it!

**Second Update:**
I switched to testing on Cori for those two tests.  Both
```
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-haswell_intel
```
and
```
PET_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.cori-knl_gnu
```
pass.

**Comment by cbegeman** [comment]
> I just put back my 3 commits that got force-removed.

Thanks. Sorry about that!

**Comment by xylar** [comment]
No worries.

**Comment by xylar** [comment]
## testing of `ocn_tracer_advection` and `*_mono`

### COMPASS

The `z_level_dev` regression test passed with Gnu compilers on my Ubuntu laptop and Intel on Anvil, with the baseline listed in my previous posts.

### E3SM

The following tests passed:
```
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_gnu.C
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel.C
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
PET_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
```

But these failed:
```
PEM_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.anvil_intel
PET_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.anvil_gnu
```

**Update:**
I rebased onto @cbegeman's latest changes and re-tested.  Both tests passed.  Again, I think the failures are something finicky with Anvil like node problems.

**Comment by cbegeman** [comment]
### Testing of ocn_diagnostics routines

**COMPASS**

The z_level_dev regression test passed with Intel compilers on grizzly against the baseline.

Baseline: 
MPAS-Dev/compass/legacy, https://github.com/MPAS-Dev/compass/commit/9f3378b278c3afd1c64c261c772271575ac426fe
with cherry-picked 
`commit 79a2aa46143f4cbecc735ffd143a1c0fd72f1567  REMOVE LATER!!! Add z-leve dev. test suite`
from https://github.com/xylar/compass/tree/add_z_level_sub_ice_shelf_2D


**E3SM**

The following tests passed:
```
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_intel.C
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu.C
PET_Ln9.T62_oQU240.GMPAS-IAF.cori-knl_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.cori-haswell_gnu
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-haswell_intel
PET_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.cori-knl_gnu
PEM_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.cori-knl_intel
```

Note: These loop limits could not be changed without introducing bfb changes in the z_level_dev regression suite.
https://github.com/cbegeman/MPAS-Model/blob/44c35746b6cd8e9b6595fd672c5583e5e64636ad/src/core_ocean/shared/mpas_ocn_diagnostics.F#L679-L689

**Comment by cbegeman** [comment]
### Testing of ocn_time_integration_split, *_vmix, *_vmix_cvmix

**COMPASS**

The z_level_dev regression test passed with Intel compilers on grizzly against the baseline.

Baseline: 
MPAS-Dev/compass/legacy, https://github.com/MPAS-Dev/compass/commit/9f3378b278c3afd1c64c261c772271575ac426fe
with cherry-picked 
`commit 79a2aa46143f4cbecc735ffd143a1c0fd72f1567  REMOVE LATER!!! Add z-leve dev. test suite`
from https://github.com/xylar/compass/tree/add_z_level_sub_ice_shelf_2D


**E3SM**
https://github.com/E3SM-Project/E3SM/commit/b3a8a9eea1598e57226018fb38956e5e91f46a12.

The following tests passed:
```
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_intel.C
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu.C
PET_Ln9.T62_oQU240.GMPAS-IAF.cori-knl_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.cori-haswell_gnu
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-haswell_intel
PET_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.cori-knl_gnu
PEM_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.cori-knl_intel
```

Note: some of the k-limits in cvmix could not be changed to `minLevel*:maxLevel*` without introducing non-bit-for-bit changes in `z_level_dev` suite. This may require further debugging for the ice-shelf test cases with `minLevelCell` not equal to 1.

**Comment by xylar** [comment]
> Note: some of the k-limits in cvmix could not be changed to minLevel*:maxLevel* without introducing non-bit-for-bit changes in z_level_dev suite. This may require further debugging for the ice-shelf test cases with minLevelCell not equal to 1.

Could you make a list of code lines where this was the case?  I'd like to investigate.

**Comment by cbegeman** [comment]
> Note: some of the k-limits in cvmix could not be changed to minLevel*:maxLevel* without introducing non-bit-for-bit changes in z_level_dev suite. This may require further debugging for the ice-shelf test cases with minLevelCell not equal to 1.

> Could you make a list of code lines where this was the case? I'd like to investigate.

It's this section:
https://github.com/cbegeman/MPAS-Model/blob/7ed4c0bd6082e744139f4e662f71fa13927fca46/src/core_ocean/shared/mpas_ocn_vmix_cvmix.F#L367-L430

**Comment by xylar** [comment]
Thanks, that does look complicated.  I'll play around with it tomorrow and see if I can make any headway.

**Comment by cbegeman** [comment]
A little thought is needed to resolve a discrepancy between `minLevelCell`,`maxLevelCell` assignments in the sub-ice-shelf 2d setup and computing `minLevelEdge*`,`minLevelVertex*` in `ocn_init_routines_compute_min_max_level`:

The problem is that we have at least one cell where `maxLevelCell` = 0 (I presume indicating an empty water column), and `minLevelCell` = 1. I imagine, though haven't verified, that `minLevelCell` is set to 1 here and not subsequently overwritten.
https://github.com/cbegeman/MPAS-Model/blob/7ed4c0bd6082e744139f4e662f71fa13927fca46/src/core_ocean/mode_init/mpas_ocn_init_sub_ice_shelf_2D.F#L248

Then `minLevelVertexTop` is also set to 1 and `layerThicknessVertex` = 0 (due to using the wrong k-level in neighboring cells), causing `ocn_diagnostic_solve_vorticity` to fail. Oh, and I should note that this is after modifying the lower loop limit to `minLevelVertexTop` on the following line:
https://github.com/cbegeman/MPAS-Model/blob/7ed4c0bd6082e744139f4e662f71fa13927fca46/src/core_ocean/shared/mpas_ocn_diagnostics.F#L679
When I set the loop limits as such for the z_level_dev test suite it failed QU240 and sub-ice-shelf-2d tests; now I think the failure could be due to a similar issue.

My inclination is to set `minLevelCell` to `nVertLevels` when the water column is empty so it's kind of analogous to `maxLevelCell` being set to 0 when the water column is empty, but I haven't thoroughly vetted this. In looking into this, I also noticed these lines, which could be problematic:
https://github.com/cbegeman/MPAS-Model/blob/7ed4c0bd6082e744139f4e662f71fa13927fca46/src/core_ocean/mode_init/mpas_ocn_init_vertical_grids.F#L518-L519

Conceptually, I'd think that when `minLevelCell`>`maxLevelCell` the water column is empty and that when `minLevelCell`=`maxLevelCell`, the water column is one cell tall.

@xylar Any thoughts since you're more familiar with these test cases and the mode_init routines?

**Comment by xylar** [comment]
> I imagine, though haven't verified, that minLevelCell is set to 1 here and not subsequently overwritten.

Yes, that's correct.

> My inclination is to set minLevelCell to nVertLevels when the water column is empty so it's kind of analogous to maxLevelCell being set to 0 when the water column is empty, but I haven't thoroughly vetted this.

I think that's nearly correct.  I think you probably want `minLevelCell = nVertLevels+1` (one-based indexing in Fortran) so `minLevelCell > maxLevelCell` even when `maxLevelCell = nVertLevels`.  I had this in the back of my mind, since this is exactly what I did in POP2x, the version of POP that had ice-shelf cavities.  But I hadn't noticed where this would be important.

I also don't think we need to set this in init mode.  I think we can modify `minLevelCell` to be `nVertLevels+1` in `ocn_init_routines_compute_min_max_level` whenever `maxLevelCell < minLevelCell`, which presumably will only happen when `maxLevelCell = 0`.

> In looking into this, I also noticed these lines, which could be problematic:

I don't see the problem with that line, other than the fact that `bottomDepth` is positive down and `ssh` is positive up, which is confusing.  That line says that the layer thickness is the water-column thickness.  In practice, I think the numerics of MPAS-Ocean don't handle situations where the water column is a single cell thick, but this is not the routine where that should be inforced so I think it remains correct.

Let's chat!

**Comment by cbegeman** [comment]
### Testing `z_level_full_cells_const_S_linear_eos`
compass: https://github.com/xylar/compass/tree/add_z_level_sub_ice_shelf_2D at commit 194ee2c 

MPAS-Model: https://github.com/cbegeman/min-level-cell-test at commit 5a65880d
- Rebased onto 3744f94c `(xylar/MPAS-Model/ocean/add_z_level_sub_ice_shelf_2D) Fix top density in landIcePressure computation`

Unless otherwise noted, `config_disable_* = .true.`
Unless otherwise noted, `config_eos_type = 'linear'`. JM EOS is the standard but the resulting density field is non-uniform and thus less convenient for testing tracer tendencies.

- [X] Thickness tendency terms: `config_disable_thick_* = .false.` and config_eos_type = 'jm'`
   - [X] Passes `check_z_level_bit_for_bit.py`
- [ ] Velocity tendency terms: `config_disable_vel_* = .false.`
   - [X] `check_z_level_bit_for_bit.py`: same level of velocity errors as when only `pgrad` is enabled
   - [ ] compare directly with case when only `pgrad` is enabled
   - [X] `config_disable_vel_all_tend = .false.`,
   `config_disable_vel_pgrad = .false.`
      - [X] `check_z_level_bit_for_bit.py`
         - `normalVelocity` not bfb due to horizontal pressure gradient force not perfectly balancing coriolis force in regions with high gradients in the depth of `minLevelCell` (threshold appears to be dz/dx ~= 60m/5km for linear EOS.
         - JM EOS also tested and yields slightly smaller `normalVelocity` fluctuations. 
- [ ] Tracer tendency terms: `config_disable_tr_* = .false.`
   - Currently segfault-ing
   - [ ] Passes `check_z_level_bit_for_bit.py`
   - [ ] `config_disable_tr_all_tend = .false.`
      - [ ] Passes `check_z_level_bit_for_bit.py`: normalVelocity not bfb

**Comment by xylar** [comment]
> Passes check_z_level_bit_for_bit.py: Currently segfault-ing

I assume the segfault is in the code, not in the python code for checking (which shouldn't even be able to segfault)?

**Comment by cbegeman** [comment]
> I assume the segfault is in the code, not in the python code for checking (which shouldn't even be able to segfault)?

Right. The way I had written it was confusing. Now edited above.

**Pull request closed without merging**